keras_nlp/utils/preset_utils.py

```diff
@@ -155,6 +160,120 @@ def save_to_preset(
+                f"Asset {asset} doesn't exist in the preset direcotry {preset}."
+            )
+    config_dir = os.path.dirname(config_path)
+    asset_dir = os.path.join(config_dir, TOKENIZER_ASSET_DIR)
+
+    tokenizer = get_tokenizer(layer)
+    if not tokenizer:
+        raise ValueError(f"Model or layer {layer} is missing tokenizer.")
+    tokenizer.load_assets(asset_dir)
+
+
+def _validate_backbone(preset):
+    config_path = get_file(preset, CONFIG_FILE)
+    with open(config_path) as config_file:
+        config = json.load(config_file)
+    # Check if backbone is deserializable.
+    keras.saving.deserialize_keras_object(config)
+
+    if config["weights"]:
+        weights_path = os.path.join(preset, config["weights"])
+        if not os.path.exists(weights_path):
+            raise FileNotFoundError(
+                f"The weights file doesn't exist in preset directory `{preset}`."
+            )
+    else:
+        raise ValueError(
+            "No weights listed in `config.json`. Make sure to use "
+            "`save_to_preset()` which adds additional data to a serialized "
+            "Keras object."
+        )
+
+
+def _validate_files(preset):
+    # TODO: check if file sizes are reasonable.
+    # TODO: validate asset files.
+    backbone_config_path = get_file(preset, CONFIG_FILE)
+    with open(backbone_config_path) as config_file:
+        backbone_config = json.load(config_file)
+    valid_files = [
+        CONFIG_FILE,
+        TOKENIZER_CONFIG_FILE,
+        "metadata.json",
+        "assets",
+        backbone_config["weights"],
+    ]
+    files = os.listdir(preset)
+    for file in files:
+        if file not in valid_files:
+            raise ValueError(f"File {file} is an unexpected file.")
+
+
+@keras_nlp_export("keras_nlp.upload_preset")
+def upload_preset(
+    uri,
+    preset,
+    allow_incomplete=False,
+):
+    """Upload a preset directory to a model hub.
+
+    Args:
+        uri: The URI identifying model to upload to.
+             URIs with format
+             `kaggle://<KAGGLE_USERNAME>/<MODEL>/<FRAMEWORK>/<VARIATION>`
+             will be uploaded to Kaggle Hub.
+        preset: The path to the local model preset directory.
+        allow_incomplete: If True, allows the upload of presets without
+                          a tokenizer configuration. Otherwise, a tokenizer
+                          is required.
+    """
+
+    # Check if preset directory exists.
+    if not os.path.exists(preset):
+        raise FileNotFoundError(f"The preset directory {preset} doesn't exist.")
+
+    _validate_backbone(preset)
+    _validate_tokenizer(preset, allow_incomplete)
+    _validate_files(preset)
+
+    if uri.startswith(KAGGLE_PREFIX):
+        kaggle_handle = uri.removeprefix(KAGGLE_PREFIX)
+        kagglehub.model_upload(kaggle_handle, preset)
+    else:
+        raise ValueError(
+            f"Unexpected URI `'{uri}'`. "
+            f"URI prefix should be one of `'{','.join([KAGGLE_PREFIX])}'`."
```

Comment:
We might want to give a entire example of a kaggle handle here, because getting all the segments is tricky (and will be different than HF).

Reply:
Added a kaggle handle example.